File: android/app/src/main/java/com/learningischange/tidbitstrivia/data/SavedQuiz.kt
package com.learningischange.tidbitstrivia.data

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlin.random.Random

/**
 * A quiz the player created and kept — the first user-authored object in Tidbits, so
 * it is a wire contract before it is a screen (docs/QUIZ-CONTRACT.md).
 *
 * A quiz stores question REFERENCES, not question text: every platform already ships
 * the corpus, so a 20-question quiz is under 1KB and costs nothing to sync, host, or
 * put in a URL. Only live-generated questions (always plain MCQ, because the corpus
 * was thin on that topic) travel inline, in the exact corpus.json row shape every
 * stack already decodes.
 *
 * Mirrors Swift SavedQuiz.swift, C# SavedQuiz.cs and JS quiz.js; pinned by
 * tools/quiz-wire/golden/quiz-v1.json.
 */
data class SavedQuiz(
    val id: String,
    val title: String,
    val topic: String,
    val creatorId: String,
    val creatorName: String,
    val createdAtMs: Long,
    val mode: String,
    val entries: List<QuizEntry>,
) {
    val questionCount: Int get() = entries.size

    /** Resolve refs in order, keeping inline questions verbatim. [lookup] returns null
     *  for an ID this build can't resolve. Never substitutes a different question — a
     *  shared quiz that quietly changes content is worse than an incomplete one. */
    /** `setLookup` comes FIRST so the trailing-lambda idiom (`quiz.resolve { ... }`)
     *  still binds to `lookup`, the one every caller needs. With the obvious ordering
     *  a trailing lambda silently became the SET lookup and every corpus ref resolved
     *  to nothing — a quiz that looked empty for no visible reason. */
    fun resolve(
        setLookup: (String, String) -> Question? = { _, _ -> null },
        lookup: (String) -> Question?,
    ): QuizResolution {
        val out = mutableListOf<Question>()
        var missing = 0
        for (e in entries) when (e) {
            is QuizEntry.Ref -> lookup(e.id)?.let { out.add(it) } ?: missing++
            // Deliberately does NOT fall back to the corpus: the corpus holds a
            // DIFFERENT question under this ID, and serving it would be the silent
            // substitution the contract forbids. Better to be one short.
            is QuizEntry.SetRef -> setLookup(e.set, e.id)?.let { out.add(it) } ?: missing++
            is QuizEntry.Inline -> out.add(e.question.toQuestion())
        }
        return QuizResolution(out, missing)
    }

    /** Keys are written in SORTED order because two devices writing the same quiz must
     *  produce byte-identical output — that is what makes the merge guard in
     *  QUIZ-CONTRACT section 4 ("created or deleted, never edited in place")
     *  checkable rather than aspirational. */
    fun toJson(): String = JSON.encodeToString(JsonElement.serializer(), toWire())

    fun toWire(): JsonObject = buildJsonObject {
        put("at", JsonPrimitive(createdAtMs))
        put("bn", JsonPrimitive(creatorName))
        put("by", JsonPrimitive(creatorId))
        put("id", JsonPrimitive(id))
        put("m", JsonPrimitive(mode))
        put("qs", buildJsonArray {
            for (e in entries) when (e) {
                is QuizEntry.Ref -> add(JsonPrimitive(e.id))
                is QuizEntry.SetRef -> add(buildJsonObject {   // keys sorted: i before s
                    put("i", JsonPrimitive(e.id))
                    put("s", JsonPrimitive(e.set))
                })
                is QuizEntry.Inline -> add(e.question.toRow())
            }
        })
        put("t", JsonPrimitive(title))
        put("tp", JsonPrimitive(topic))
        put("v", JsonPrimitive(1))
    }

    companion object {
        /** Crockford-style, with 0/o, 1/l/i and u removed so an ID read aloud in a pub
         *  or typed off a projector is unambiguous, and a random ID can't spell
         *  something unfortunate. 30 chars: 30^10 ~= 5.9e14. */
        const val ID_ALPHABET = "23456789abcdefghjkmnpqrstvwxyz"
        const val ID_LENGTH = 10

        /** The modes a created quiz can be played as. Deliberately a SUBSET: a saved
         *  quiz deals a FIXED set of questions, so any mode that draws its own
         *  (daily, marathon, weakSpot) would silently ignore the very questions the
         *  quiz exists to preserve. */
        val PLAYABLE_MODES = listOf("mix", "classic", "timeAttack", "survival", "stake")

        fun modeLabel(m: String): String = when (m) {
            "classic" -> "Classic"
            "timeAttack" -> "Time Attack"
            "survival" -> "Survival"
            "stake" -> "Stake"
            else -> "Mixed shapes"
        }

        /** An unknown mode from a newer build must still PLAY rather than refuse —
         *  these objects outlive the version that wrote them (QUIZ-CONTRACT §6). */
        fun playableMode(m: String): String = if (m in PLAYABLE_MODES) m else "mix"

        /** Below this a quiz isn't worth playing; above it we play and say so. */
        const val MINIMUM_PLAYABLE = 3

        private val JSON = Json { ignoreUnknownKeys = true }

        /** Random, never derived from content: two people who both make a "Jazz" quiz
         *  must get different IDs, and an ID must not leak what is inside it. */
        fun makeId(rng: Random = Random.Default): String =
            (1..ID_LENGTH).map { ID_ALPHABET[rng.nextInt(ID_ALPHABET.length)] }.joinToString("")

        /** Titles ride in share cards and list rows, so they are trimmed and capped
         *  rather than rejected — a long paste should still save. */
        fun cleanTitle(raw: String?): String {
            val t = (raw ?: "").trim()
            if (t.isEmpty()) return "Untitled quiz"
            return if (t.length <= 60) t else t.substring(0, 60)
        }

        /** Live Wikipedia generation is the only source that isn't addressable by ID
         *  from a bundled file, so it is the only thing a quiz has to carry inline. */
        fun isLiveGenerated(q: Question): Boolean = q.id.startsWith("live:")

        /** Which bundled set a question came from, or null for a plain corpus row.
         *  Derived from the question's SHAPE rather than threaded through from the
         *  call site, so it stays correct no matter which surface built the set. */
        fun bundledSetName(q: Question): String? = when {
            q.imageUrl != null -> "picture"
            q.closest != null -> "closest"
            q.ordering != null -> "order"
            q.matching != null -> "match"
            q.accepted != null -> "typeanswer"
            q.enumerate != null -> "enumerate"
            q.id.startsWith("tot:") -> "thisorthat"
            q.id.startsWith("odd:") -> "oddoneout"
            else -> null
        }

        fun from(
            questions: List<Question>,
            topic: String,
            creatorId: String,
            creatorName: String,
            title: String? = null,
            mode: String = "mix",
            id: String? = null,
            createdAtMs: Long = System.currentTimeMillis(),
        ) = SavedQuiz(
            id = id ?: makeId(),
            title = cleanTitle(title ?: topic),
            topic = topic,
            creatorId = creatorId,
            creatorName = creatorName,
            createdAtMs = createdAtMs,
            mode = mode,
            entries = questions.map {
                when {
                    isLiveGenerated(it) -> QuizEntry.Inline(InlineQuestion.of(it))
                    else -> bundledSetName(it)
                        ?.let { set -> QuizEntry.SetRef(set, it.id) } ?: QuizEntry.Ref(it.id)
                }
            },
        )

        /** Lenient by contract: unknown keys are ignored and a malformed entry is
         *  skipped rather than failing the whole quiz, because these objects outlive
         *  the app version that wrote them. */
        fun fromJson(text: String): SavedQuiz? = try {
            fromWire(JSON.parseToJsonElement(text).jsonObject)
        } catch (_: Exception) {
            null
        }

        fun fromWire(w: JsonObject): SavedQuiz? {
            val id = str(w, "id") ?: return null
            if (id.isEmpty()) return null
            val by = str(w, "by") ?: return null
            val qs = (w["qs"] as? JsonArray) ?: return null
            val entries = qs.mapNotNull { raw ->
                val prim = raw as? JsonPrimitive
                if (prim != null && prim.isString) {
                    prim.content.takeIf { it.isNotEmpty() }?.let { QuizEntry.Ref(it) }
                } else if (raw is JsonObject) {
                    val set = (raw["s"] as? JsonPrimitive)?.takeIf { it.isString }?.content
                    val qid = (raw["i"] as? JsonPrimitive)?.takeIf { it.isString }?.content
                    if (!set.isNullOrEmpty() && !qid.isNullOrEmpty()) QuizEntry.SetRef(set, qid) else null
                } else {
                    (raw as? JsonArray)?.let { InlineQuestion.fromRow(it) }?.let { QuizEntry.Inline(it) }
                }
            }
            return SavedQuiz(
                id = id,
                title = cleanTitle(str(w, "t")),
                topic = str(w, "tp") ?: "",
                creatorId = by,
                creatorName = str(w, "bn") ?: "",
                createdAtMs = (w["at"] as? JsonPrimitive)?.longOrNull ?: 0L,
                mode = str(w, "m") ?: "mix",
                entries = entries,
            )
        }

        private fun str(o: JsonObject, key: String): String? =
            (o[key] as? JsonPrimitive)?.takeIf { it.isString }?.content
    }
}

/** One entry in a quiz's ordered question list: a REF (corpus/bundled-set question ID)
 *  or an INLINE question in corpus.json row shape. */
sealed interface QuizEntry {
    data class Ref(val id: String) : QuizEntry

    /** A BUNDLED-SET question, carrying which set it came from. Exists because a bare
     *  ID is genuinely ambiguous: the bundled sets share the corpus `src:` namespace,
     *  and 166 of 200 sampled Picture ID rows have an ID that ALSO exists in the corpus
     *  as a different question shape, so resolving corpus-first served a text question
     *  in place of a saved picture question. */
    data class SetRef(val set: String, val id: String) : QuizEntry

    data class Inline(val question: InlineQuestion) : QuizEntry
}

/** What a reader got back after resolving refs. Refs go missing legitimately (older
 *  build, retired row), so this reports the shortfall instead of hiding it. */
data class QuizResolution(val questions: List<Question>, val missing: Int) {
    val isPlayable: Boolean get() = questions.size >= SavedQuiz.MINIMUM_PLAYABLE
    val isComplete: Boolean get() = missing == 0
}

/** A live-generated MCQ carried inside a quiz, in the corpus.json row shape:
 *  [id, prompt, [o0,o1,o2,o3], correctIndex, category, difficulty, explanation,
 *  sourceTitle, sourceURL]. Reusing the corpus row is the whole point — every stack
 *  already decodes it, so the six implementations cannot drift. */
data class InlineQuestion(
    val id: String,
    val prompt: String,
    val options: List<String>,
    val correctIndex: Int,
    val categoryId: String,
    val difficulty: Int,
    val explanation: String,
    val sourceTitle: String,
    val sourceUrl: String,
) {
    fun toRow(): JsonArray = buildJsonArray {
        add(JsonPrimitive(id))
        add(JsonPrimitive(prompt))
        add(buildJsonArray { options.forEach { add(JsonPrimitive(it)) } })
        add(JsonPrimitive(correctIndex))
        add(JsonPrimitive(categoryId))
        add(JsonPrimitive(difficulty))
        add(JsonPrimitive(explanation))
        add(JsonPrimitive(sourceTitle))
        add(JsonPrimitive(sourceUrl))
    }

    fun toQuestion() = Question(
        id = id, prompt = prompt, options = options, correctIndex = correctIndex,
        categoryId = categoryId, difficulty = difficulty, explanation = explanation,
        sourceTitle = sourceTitle, sourceUrl = sourceUrl,
    )

    companion object {
        fun of(q: Question) = InlineQuestion(
            q.id, q.prompt, q.options, q.correctIndex, q.categoryId,
            q.difficulty, q.explanation, q.sourceTitle, q.sourceUrl,
        )

        fun fromRow(row: JsonArray): InlineQuestion? {
            if (row.size < 9) return null
            val id = strAt(row, 0) ?: return null
            val prompt = strAt(row, 1) ?: return null
            val opts = (row[2] as? JsonArray)?.takeIf { it.size == 4 }
                ?.mapNotNull { (it as? JsonPrimitive)?.contentOrNull } ?: return null
            if (opts.size != 4) return null
            val correct = (row[3] as? JsonPrimitive)?.intOrNull ?: return null
            val cat = strAt(row, 4) ?: return null
            val diff = (row[5] as? JsonPrimitive)?.intOrNull ?: return null
            val expl = strAt(row, 6) ?: return null
            val title = strAt(row, 7) ?: return null
            return InlineQuestion(id, prompt, opts, correct, cat, diff, expl, title,
                strAt(row, 8) ?: "")
        }

        private fun strAt(a: JsonArray, i: Int): String? =
            (a.getOrNull(i) as? JsonPrimitive)?.takeIf { it.isString }?.content
    }
}

File: windows/Tidbits.Core/Models/SavedQuiz.cs
using System.Text;
using System.Text.Encodings.Web;
using System.Text.Json;

namespace Tidbits.Core.Models;

/// <summary>
/// One entry in a quiz's ordered question list: a CORPUS ref (a bare string on the
/// wire), a BUNDLED-SET ref (carrying which set it came from), or an INLINE question.
///
/// SetRef exists because a bare ID is genuinely ambiguous: the bundled sets share the
/// corpus "src:" namespace, and 166 of 200 sampled Picture ID rows have an ID that
/// ALSO exists in the corpus as a different question shape. Resolving corpus-first
/// therefore served a text question in place of a saved picture question.
/// </summary>
public readonly record struct QuizEntry(string? Ref, string? Set, InlineQuestion? Inline)
{
    public static QuizEntry OfRef(string id) => new(id, null, null);
    public static QuizEntry OfSetRef(string set, string id) => new(id, set, null);
    public static QuizEntry OfInline(InlineQuestion q) => new(null, null, q);
    public bool IsRef => Ref is not null && Set is null;
    public bool IsSetRef => Set is not null;
}

/// <summary>
/// A live-generated MCQ carried inside a quiz, in the corpus.json row shape:
/// [id, prompt, [o0,o1,o2,o3], correctIndex, category, difficulty, explanation,
/// sourceTitle, sourceURL]. Reusing the corpus row is the whole point — every stack
/// already decodes it, so the six implementations cannot drift.
/// </summary>
public sealed record InlineQuestion(
    string Id, string Prompt, IReadOnlyList<string> Options, int CorrectIndex,
    string CategoryId, int Difficulty, string Explanation, string SourceTitle, string SourceUrl)
{
    public static InlineQuestion? FromRow(JsonElement row)
    {
        if (row.ValueKind != JsonValueKind.Array || row.GetArrayLength() < 9) return null;
        var a = row.EnumerateArray().ToArray();
        if (a[0].ValueKind != JsonValueKind.String || a[1].ValueKind != JsonValueKind.String) return null;
        if (a[2].ValueKind != JsonValueKind.Array || a[2].GetArrayLength() != 4) return null;
        if (a[3].ValueKind != JsonValueKind.Number || a[5].ValueKind != JsonValueKind.Number) return null;
        if (a[4].ValueKind != JsonValueKind.String || a[6].ValueKind != JsonValueKind.String
            || a[7].ValueKind != JsonValueKind.String) return null;
        return new InlineQuestion(
            a[0].GetString()!, a[1].GetString()!,
            a[2].EnumerateArray().Select(o => o.GetString() ?? "").ToList(),
            a[3].GetInt32(), a[4].GetString()!, a[5].GetInt32(),
            a[6].GetString()!, a[7].GetString()!,
            a[8].ValueKind == JsonValueKind.String ? a[8].GetString()! : "");
    }

    public Question ToQuestion() => new()
    {
        Id = Id, Prompt = Prompt, Options = Options.ToList(), CorrectIndex = CorrectIndex,
        CategoryId = CategoryId, Difficulty = Difficulty, Explanation = Explanation,
        SourceTitle = SourceTitle, SourceUrl = SourceUrl,
        TemplateId = Id.Split(':').FirstOrDefault() ?? "live",
    };
}

/// <summary>
/// What a reader got back after resolving refs against the local corpus. Refs go
/// missing legitimately (older build, retired row), so this reports the shortfall
/// instead of hiding it.
/// </summary>
public sealed record QuizResolution(IReadOnlyList<Question> Questions, int Missing)
{
    public bool IsPlayable => Questions.Count >= SavedQuiz.MinimumPlayable;
    public bool IsComplete => Missing == 0;
}

/// <summary>
/// A quiz the player created and kept — the first user-authored object in Tidbits,
/// so it is a wire contract before it is a screen (docs/QUIZ-CONTRACT.md).
///
/// A quiz stores question REFERENCES, not question text: every platform already ships
/// the corpus, so a 20-question quiz is under 1KB and costs nothing to sync, host, or
/// put in a URL. Only live-generated questions (always plain MCQ) travel inline.
///
/// Mirrors Swift SavedQuiz.swift, Kotlin SavedQuiz.kt and JS quiz.js; pinned by
/// tools/quiz-wire/golden/quiz-v1.json.
/// </summary>
public sealed class SavedQuiz
{
    /// Crockford-style, with 0/o, 1/l/i and u removed so an ID read aloud in a pub or
    /// typed off a projector is unambiguous, and a random ID can't spell something
    /// unfortunate. 30 chars: 30^10 ~= 5.9e14.
    public const string IdAlphabet = "23456789abcdefghjkmnpqrstvwxyz";
    public const int IdLength = 10;

    /// Below this a quiz isn't worth playing; above it we play and say so.
    public const int MinimumPlayable = 3;

    public required string Id { get; init; }
    public required string Title { get; set; }
    public required string Topic { get; init; }
    /// Settable because publishing stamps the authenticated uid: the RTDB rules only
    /// let `by === auth.uid` overwrite a quiz, so a locally-created quiz saved as
    /// "local" must carry the real uid once it is shared.
    public required string CreatorId { get; set; }
    public required string CreatorName { get; init; }
    public required long CreatedAtMs { get; init; }
    public required string Mode { get; set; }
    public required List<QuizEntry> Entries { get; set; }

    public int QuestionCount => Entries.Count;

    /// Random, never derived from content: two people who both make a "Jazz" quiz must
    /// get different IDs, and an ID must not leak what is inside it.
    public static string MakeId(Random? rng = null)
    {
        rng ??= Random.Shared;
        var sb = new StringBuilder(IdLength);
        for (int i = 0; i < IdLength; i++) sb.Append(IdAlphabet[rng.Next(IdAlphabet.Length)]);
        return sb.ToString();
    }

    /// Titles ride in share cards and list rows, so they are trimmed and capped rather
    /// than rejected — a long paste should still save.
    public static string CleanTitle(string? raw)
    {
        var t = (raw ?? "").Trim();
        if (t.Length == 0) return "Untitled quiz";
        return t.Length <= 60 ? t : t[..60];
    }

    /// Live Wikipedia generation is the only source that isn't addressable by ID from
    /// a bundled file, so it is the only thing a quiz has to carry inline.
    public static bool IsLiveGenerated(Question q)
        => q.Id.StartsWith("live:") || q.TemplateId == "live";

    /// Which bundled set a question came from, or null for a plain corpus row.
    /// Derived from the question's SHAPE rather than threaded through from the call
    /// site, so it stays correct no matter which surface built the set.
    public static string? BundledSetName(Question q)
    {
        if (q.ImageUrl is not null) return "picture";
        if (q.Closest is not null) return "closest";
        if (q.Ordering is not null) return "order";
        if (q.Matching is not null) return "match";
        if (q.Accepted is not null) return "typeanswer";
        if (q.Enumerate is not null) return "enumerate";
        if (q.Id.StartsWith("tot:")) return "thisorthat";
        if (q.Id.StartsWith("odd:")) return "oddoneout";
        return null;
    }

    public static SavedQuiz From(IEnumerable<Question> questions, string topic, string creatorId,
                                 string creatorName, string? title = null, string mode = "mix",
                                 string? id = null, long? createdAtMs = null) => new()
    {
        Id = id ?? MakeId(),
        Title = CleanTitle(title ?? topic),
        Topic = topic,
        CreatorId = creatorId,
        CreatorName = creatorName,
        CreatedAtMs = createdAtMs ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
        Mode = mode,
        Entries = questions.Select(q =>
        {
            if (IsLiveGenerated(q))
                return QuizEntry.OfInline(new InlineQuestion(q.Id, q.Prompt, q.Options, q.CorrectIndex,
                    q.CategoryId, q.Difficulty, q.Explanation, q.SourceTitle, q.SourceUrl ?? ""));
            var set = BundledSetName(q);
            return set is not null ? QuizEntry.OfSetRef(set, q.Id) : QuizEntry.OfRef(q.Id);
        }).ToList(),
    };

    /// Resolve in order, keeping inline questions verbatim. <paramref name="lookup"/>
    /// returns null for an ID this build can't resolve. Never substitutes a different
    /// question — a shared quiz that quietly changes content is worse than one that
    /// admits it is incomplete.
    public QuizResolution Resolve(Func<string, Question?> lookup,
                                  Func<string, string, Question?>? setLookup = null)
    {
        var outQ = new List<Question>();
        int missing = 0;
        foreach (var e in Entries)
        {
            if (e.IsSetRef)
            {
                // Deliberately does NOT fall back to the corpus: the corpus holds a
                // DIFFERENT question under this ID, and serving it would be the silent
                // substitution the contract forbids. Better to be one short.
                var q = setLookup?.Invoke(e.Set!, e.Ref!);
                if (q is not null) outQ.Add(q); else missing++;
            }
            else if (e.IsRef)
            {
                var q = lookup(e.Ref!);
                if (q is not null) outQ.Add(q); else missing++;
            }
            else outQ.Add(e.Inline!.ToQuestion());
        }
        return new QuizResolution(outQ, missing);
    }

    // MARK: Wire codec (docs/QUIZ-CONTRACT.md §2)

    /// Keys are written in SORTED order and slashes/quotes are left unescaped, because
    /// two devices writing the same quiz must produce byte-identical output. Both are
    /// load-bearing for cross-stack identity, not cosmetics: System.Text.Json escapes
    /// an apostrophe to ' by default, so a quiz about "Destiny's Child" would
    /// differ byte-for-byte from every other stack while decoding to the same object.
    private static readonly JsonWriterOptions WriterOptions = new()
    {
        Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
        Indented = false,
    };

    public string ToJson()
    {
        using var stream = new MemoryStream();
        using (var w = new Utf8JsonWriter(stream, WriterOptions))
        {
            w.WriteStartObject();
            w.WriteNumber("at", CreatedAtMs);
            w.WriteString("bn", CreatorName);
            w.WriteString("by", CreatorId);
            w.WriteString("id", Id);
            w.WriteString("m", Mode);
            w.WriteStartArray("qs");
            foreach (var e in Entries)
            {
                if (e.IsSetRef)
                {
                    w.WriteStartObject();          // keys sorted: i before s
                    w.WriteString("i", e.Ref!);
                    w.WriteString("s", e.Set!);
                    w.WriteEndObject();
                    continue;
                }
                if (e.IsRef) { w.WriteStringValue(e.Ref!); continue; }
                var i = e.Inline!;
                w.WriteStartArray();
                w.WriteStringValue(i.Id);
                w.WriteStringValue(i.Prompt);
                w.WriteStartArray();
                foreach (var o in i.Options) w.WriteStringValue(o);
                w.WriteEndArray();
                w.WriteNumberValue(i.CorrectIndex);
                w.WriteStringValue(i.CategoryId);
                w.WriteNumberValue(i.Difficulty);
                w.WriteStringValue(i.Explanation);
                w.WriteStringValue(i.SourceTitle);
                w.WriteStringValue(i.SourceUrl);
                w.WriteEndArray();
            }
            w.WriteEndArray();
            w.WriteString("t", Title);
            w.WriteString("tp", Topic);
            w.WriteNumber("v", 1);
            w.WriteEndObject();
        }
        return Encoding.UTF8.GetString(stream.ToArray());
    }

    /// Lenient by contract: unknown keys are ignored and a malformed entry is skipped
    /// rather than failing the whole quiz, because these objects outlive the app
    /// version that wrote them.
    public static SavedQuiz? FromJson(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return null;
            if (!root.TryGetProperty("id", out var idEl) || idEl.ValueKind != JsonValueKind.String) return null;
            var id = idEl.GetString()!;
            if (id.Length == 0) return null;
            if (!root.TryGetProperty("by", out var byEl) || byEl.ValueKind != JsonValueKind.String) return null;
            if (!root.TryGetProperty("qs", out var qsEl) || qsEl.ValueKind != JsonValueKind.Array) return null;

            var entries = new List<QuizEntry>();
            foreach (var raw in qsEl.EnumerateArray())
            {
                if (raw.ValueKind == JsonValueKind.String)
                {
                    var s = raw.GetString()!;
                    if (s.Length > 0) entries.Add(QuizEntry.OfRef(s));
                }
                else if (raw.ValueKind == JsonValueKind.Object)
                {
                    if (raw.TryGetProperty("s", out var setEl) && setEl.ValueKind == JsonValueKind.String
                        && raw.TryGetProperty("i", out var idEl2) && idEl2.ValueKind == JsonValueKind.String)
                    {
                        var set = setEl.GetString()!;
                        var qid = idEl2.GetString()!;
                        if (set.Length > 0 && qid.Length > 0) entries.Add(QuizEntry.OfSetRef(set, qid));
                    }
                }
                else
                {
                    var inline = InlineQuestion.FromRow(raw);
                    if (inline is not null) entries.Add(QuizEntry.OfInline(inline));
                }
            }

            return new SavedQuiz
            {
                Id = id,
                Title = CleanTitle(Str(root, "t")),
                Topic = Str(root, "tp") ?? "",
                CreatorId = byEl.GetString()!,
                CreatorName = Str(root, "bn") ?? "",
                CreatedAtMs = root.TryGetProperty("at", out var at) && at.ValueKind == JsonValueKind.Number
                    ? at.GetInt64() : 0,
                Mode = Str(root, "m") ?? "mix",
                Entries = entries,
            };
        }
        catch (JsonException) { return null; }
    }

    private static string? Str(JsonElement o, string key)
        => o.TryGetProperty(key, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;
}
